File: apps/api/src/modules/llm/cost/session-cost.tracker.ts
/**
 * 会话级成本面板（P0 配套）
 *
 * 入口：
 *  - startSession(interviewId)   - 面试开始
 *  - recordLlmCall(interviewId, data) - 每次 LLM 调用埋点
 *  - endSession(interviewId)    - 面试结束
 *
 * 存储：Prisma SessionCost + Redis 实时 counter（防 DB 抖动）
 *
 * 响应字段（GET /api/session/:id/cost）：
 *  {
 *    sessionId, totalTokens, llmCalls, promptCacheHits, semanticCacheHits,
 *    cacheSavedTokens, retryRate, fallbackRate, estimatedCostCny, durationMs
 *  }
 */

import { Injectable, Logger, OnModuleInit } from '@nestjs/common';
import { PrismaService } from '../../../infra/prisma/prisma.service';
import { RedisService } from '../../../infra/redis/redis.service';

const REDIS_KEY_PREFIX = 'session_cost:';
const REDIS_FLUSH_EVERY = 5; // 每 5 次调用刷一次 DB（防抖）

export interface LlmCallMetric {
  interviewId: string;
  provider: string;
  model: string;
  promptTokens: number;
  completionTokens: number;
  cachedTokens: number; // prompt cache 命中 token
  cacheHit: boolean; // semantic cache 命中
  isRetry: boolean;
  isFallback: boolean;
  isError?: boolean;
  durationMs: number;
}

@Injectable()
export class SessionCostTracker implements OnModuleInit {
  private readonly logger = new Logger(SessionCostTracker.name);
  /** 内存 buffer：interviewId → 未刷盘计数 */
  private buffer: Map<string, number> = new Map();

  constructor(private prisma: PrismaService, private redis: RedisService) {}

  async onModuleInit() {
    // 启动时从 Redis 恢复 buffer（用于崩溃恢复）
    // 实际可省略：buffer 是临时累加，崩溃丢失可接受
  }

  async startSession(interviewId: string): Promise<void> {
    // 幂等创建
    await this.prisma.sessionCost.upsert({
      where: { interviewId },
      create: { interviewId, startedAt: new Date() },
      update: { startedAt: new Date() },
    });
    // 清 Redis counter
    await this.redis.del(this.redisKey(interviewId));
  }

  /**
   * 每次 LLM 调用后调用
   * 写入 Redis counter（实时读）+ 累计 N 次后刷 Prisma
   */
  async recordLlmCall(m: LlmCallMetric): Promise<void> {
    // 2026-06-24 防御性修复：interviewId='unknown' / 空值直接 skip
    // （不写 Redis、不累计 buffer、不触发 FK flush）。
    // 调用方已修（dynamic-task-queue / generateReport 显式传 ctx.sessionId），
    // 这层兜底防止将来再有漏传。
    if (!m.interviewId || m.interviewId === 'unknown' || m.interviewId === 'anonymous') {
      this.logger.warn(`[recordLlmCall] skip invalid interviewId=${m.interviewId} (cost tracking only)`);
      return;
    }
    const key = this.redisKey(m.interviewId);
    const redis = this.redis.getClient();

    // Redis HINCRBY pipeline - 实时 counter
    const pipe = redis.pipeline();
    pipe.hincrby(key, 'llmCalls', 1);
    pipe.hincrby(key, 'totalPromptTokens', m.promptTokens);
    pipe.hincrby(key, 'totalCompletionTokens', m.completionTokens);
    pipe.hincrby(key, 'totalTokens', m.promptTokens + m.completionTokens);
    if (m.cachedTokens > 0) {
      pipe.hincrby(key, 'promptCacheHits', 1);
      pipe.hincrby(key, 'cachedTokens', m.cachedTokens);
      // 节省的 token ≈ 命中部分按 input 价 + 实际按 cache 折扣计 → 差值即节省
      pipe.hincrby(key, 'cacheSavedTokens', m.cachedTokens);
    } else if (!m.cacheHit) {
      // 既不是 prompt cache 命中，也不是 semantic cache 命中 → 算 prompt cache miss
      // 但只有同 user 的稳定前缀调用才算 miss；流式首包不算
      // 简化：除 semantic hit 外的都算 miss
      if (m.provider === 'qwen' || m.provider === 'deepseek') {
        pipe.hincrby(key, 'promptCacheMisses', 1);
      }
    }
    if (m.cacheHit) pipe.hincrby(key, 'semanticCacheHits', 1);
    if (m.isRetry) pipe.hincrby(key, 'retries', 1);
    if (m.isFallback) pipe.hincrby(key, 'fallbacks', 1);
    if (m.isError) pipe.hincrby(key, 'errors', 1);
    await pipe.exec();

    // buffer 累加，5 次刷一次 DB
    const buf = (this.buffer.get(m.interviewId) || 0) + 1;
    this.buffer.set(m.interviewId, buf);
    if (buf >= REDIS_FLUSH_EVERY) {
      await this.flushToDb(m.interviewId);
      this.buffer.set(m.interviewId, 0);
    }
  }

  async endSession(interviewId: string): Promise<void> {
    await this.flushToDb(interviewId);
    await this.prisma.sessionCost.update({
      where: { interviewId },
      data: { endedAt: new Date() },
    });
  }

  /** 强刷：用于 GET endpoint 调用时拿到最新值 */
  async flushToDb(interviewId: string): Promise<void> {
    this.logger.warn(`[FLUSH] start interviewId=${interviewId}`);
    // 2026-06-24 防御性修复：interviewId='unknown' 或空值直接 return，
    // 避免对 session_costs 表触发 FK 违反 + 抛错中断主流程。
    // 这是兜底：调用方（dynamic-task-queue / generateReport）现在已修，
    // 不会漏传 interviewId；但如果将来又有调用方漏传，不至于把整个 end
    // 流程炸成"生成报告失败"占位报告。
    if (!interviewId || interviewId === 'unknown' || interviewId === 'anonymous') {
      this.logger.warn(`[FLUSH] skip invalid interviewId=${interviewId} (cost tracking only)`);
      return;
    }
    const key = this.redisKey(interviewId);
    const raw = await this.redis.getClient().hgetall(key);
    if (!raw || Object.keys(raw).length === 0) return;

    // 计算 cost（按 provider 单价，可从 env 读）
    const inputPrice = parseFloat(process.env.QWEN_INPUT_PRICE || '0.004'); // 元/1k tokens
    const outputPrice = parseFloat(process.env.QWEN_OUTPUT_PRICE || '0.012');
    const cacheDiscount = 0.4; // Qwen context cache 折扣

    const totalPrompt = Number(raw.totalPromptTokens || 0);
    const totalCompletion = Number(raw.totalCompletionTokens || 0);
    const cachedTokens = Number(raw.cachedTokens || 0);
    const uncachedPrompt = totalPrompt - cachedTokens;
    const cost =
      (uncachedPrompt / 1000) * inputPrice +
      (cachedTokens / 1000) * inputPrice * cacheDiscount +
      (totalCompletion / 1000) * outputPrice;

    await this.prisma.sessionCost.upsert({
      where: { interviewId },
      create: {
        interviewId,
        llmCalls: Number(raw.llmCalls || 0),
        totalPromptTokens: totalPrompt,
        totalCompletionTokens: totalCompletion,
        totalTokens: totalPrompt + totalCompletion, // 冗余字段，写入时计算
        promptCacheHits: Number(raw.promptCacheHits || 0),
        promptCacheMisses: Number(raw.promptCacheMisses || 0),
        cachedTokens,
        semanticCacheHits: Number(raw.semanticCacheHits || 0),
        semanticCacheMisses: Number(raw.semanticCacheMisses || 0),
        cacheSavedTokens: Number(raw.cacheSavedTokens || 0),
        retries: Number(raw.retries || 0),
        fallbacks: Number(raw.fallbacks || 0),
        errors: Number(raw.errors || 0),
        inputCostPer1k: inputPrice,
        outputCostPer1k: outputPrice,
        cacheDiscount,
        estimatedCostCny: cost,
      },
      update: {
        llmCalls: Number(raw.llmCalls || 0),
        totalPromptTokens: totalPrompt,
        totalCompletionTokens: totalCompletion,
        totalTokens: totalPrompt + totalCompletion, // 同步刷新
        promptCacheHits: Number(raw.promptCacheHits || 0),
        promptCacheMisses: Number(raw.promptCacheMisses || 0),
        cachedTokens,
        semanticCacheHits: Number(raw.semanticCacheHits || 0),
        semanticCacheMisses: Number(raw.semanticCacheMisses || 0),
        cacheSavedTokens: Number(raw.cacheSavedTokens || 0),
        retries: Number(raw.retries || 0),
        fallbacks: Number(raw.fallbacks || 0),
        errors: Number(raw.errors || 0),
        estimatedCostCny: cost,
      },
    }).then(() => {
      this.logger.warn(`[FLUSH] OK interviewId=${interviewId}`);
    }).catch((e) => {
      this.logger.error(`[FLUSH] FAIL interviewId=${interviewId} err=${e.message} code=${e.code} meta=${JSON.stringify(e.meta)}`);
      throw e;
    });
  }

  /**
   * GET /api/session/:id/cost 用的快路径
   * 先 flush 再读 Prisma，1s 内返回
   */
  async getCostPanel(interviewId: string) {
    await this.flushToDb(interviewId);
    const row = await this.prisma.sessionCost.findUnique({ where: { interviewId } });
    if (!row) {
      return {
        sessionId: interviewId,
        totalTokens: 0,
        llmCalls: 0,
        promptCacheHits: 0,
        promptCacheMisses: 0,
        promptCacheHitRate: 0,
        semanticCacheHits: 0,
        semanticCacheMisses: 0,
        semanticCacheHitRate: 0,
        cacheSavedTokens: 0,
        retryRate: 0,
        fallbackRate: 0,
        estimatedCostCny: 0,
        durationMs: 0,
      };
    }

    const promptTotal = row.promptCacheHits + row.promptCacheMisses;
    const semanticTotal = row.semanticCacheHits + row.semanticCacheMisses;
    const durationMs = row.endedAt
      ? row.endedAt.getTime() - row.startedAt.getTime()
      : Date.now() - row.startedAt.getTime();

    return {
      sessionId: interviewId,
      totalTokens: row.totalTokens,
      llmCalls: row.llmCalls,
      promptCacheHits: row.promptCacheHits,
      promptCacheMisses: row.promptCacheMisses,
      promptCacheHitRate: promptTotal > 0 ? +(row.promptCacheHits / promptTotal).toFixed(4) : 0,
      semanticCacheHits: row.semanticCacheHits,
      semanticCacheMisses: row.semanticCacheMisses,
      semanticCacheHitRate: semanticTotal > 0 ? +(row.semanticCacheHits / semanticTotal).toFixed(4) : 0,
      cacheSavedTokens: row.cacheSavedTokens,
      retryRate: row.llmCalls > 0 ? +(row.retries / row.llmCalls).toFixed(4) : 0,
      fallbackRate: row.llmCalls > 0 ? +(row.fallbacks / row.llmCalls).toFixed(4) : 0,
      errors: row.errors,
      estimatedCostCny: +row.estimatedCostCny.toFixed(4),
      durationMs,
    };
  }

  private redisKey(interviewId: string): string {
    return `${REDIS_KEY_PREFIX}${interviewId}`;
  }
}

File: apps/py/src/interview_agent/modules/llm/cost/session_cost_tracker.py
"""SessionCostTracker — 与 NestJS SessionCostTracker 像素级对齐。

存储：Prisma SessionCost + Redis 实时 counter
- Redis HINCRBY pipeline（实时读，< 100ms）
- 每 5 次刷一次 DB（防抖）
- GET /api/session/:id/cost 先 flush 再读
"""
import json
import logging
import os
from typing import Any

from sqlalchemy import select

from interview_agent.infra.db import async_session_factory
from interview_agent.infra.models import SessionCost
from interview_agent.infra.redis_client import get_redis

logger = logging.getLogger(__name__)

REDIS_KEY_PREFIX = "session_cost:"
REDIS_FLUSH_EVERY = 5


class SessionCostTracker:
    """会话级成本面板单例。"""

    _instance: "SessionCostTracker | None" = None
    _buffer: dict[str, int] = {}

    def __init__(self):
        pass

    @classmethod
    def instance(cls) -> "SessionCostTracker":
        if cls._instance is None:
            cls._instance = cls()
        return cls._instance

    @staticmethod
    def _redis_key(interview_id: str) -> str:
        return f"{REDIS_KEY_PREFIX}{interview_id}"

    async def start_session(self, interview_id: str) -> None:
        """启动 session：幂等创建 SessionCost + 清 Redis counter。"""
        async with async_session_factory() as session:
            existing = await session.get(SessionCost, interview_id) if False else None
            # 实际上 SessionCost 用 interviewId 唯一，所以 upsert：
            stmt = select(SessionCost).where(SessionCost.interview_id == interview_id)
            result = await session.execute(stmt)
            row = result.scalar_one_or_none()
            if row is None:
                import secrets
                session.add(SessionCost(
                    id=f"c{secrets.token_hex(12)}",
                    interview_id=interview_id,
                ))
                await session.commit()
        redis = get_redis()
        await redis.delete(self._redis_key(interview_id))

    async def record_llm_call(self, metric: dict[str, Any]) -> None:
        """每次 LLM 调用埋点。"""
        interview_id = metric.get("interviewId", "")
        # 防御性：invalid interviewId skip
        if not interview_id or interview_id in ("unknown", "anonymous"):
            logger.warning(
                f"[recordLlmCall] skip invalid interviewId={interview_id} (cost tracking only)"
            )
            return

        key = self._redis_key(interview_id)
        redis = get_redis()

        # Redis HINCRBY pipeline
        pipe = redis.pipeline()
        pipe.hincrby(key, "llmCalls", 1)
        pipe.hincrby(key, "totalPromptTokens", metric.get("promptTokens", 0))
        pipe.hincrby(key, "totalCompletionTokens", metric.get("completionTokens", 0))
        pipe.hincrby(key, "totalTokens",
                     metric.get("promptTokens", 0) + metric.get("completionTokens", 0))
        if metric.get("cachedTokens", 0) > 0:
            pipe.hincrby(key, "promptCacheHits", 1)
            pipe.hincrby(key, "cachedTokens", metric["cachedTokens"])
            pipe.hincrby(key, "cacheSavedTokens", metric["cachedTokens"])
        elif not metric.get("cacheHit", False):
            if metric.get("provider") in ("qwen", "deepseek"):
                pipe.hincrby(key, "promptCacheMisses", 1)
        if metric.get("cacheHit", False):
            pipe.hincrby(key, "semanticCacheHits", 1)
        if metric.get("isRetry", False):
            pipe.hincrby(key, "retries", 1)
        if metric.get("isFallback", False):
            pipe.hincrby(key, "fallbacks", 1)
        if metric.get("isError", False):
            pipe.hincrby(key, "errors", 1)
        await pipe.execute()

        # Buffer 累加，每 5 次刷盘
        buf = self._buffer.get(interview_id, 0) + 1
        self._buffer[interview_id] = buf
        if buf >= REDIS_FLUSH_EVERY:
            await self.flush_to_db(interview_id)
            self._buffer[interview_id] = 0

    async def end_session(self, interview_id: str) -> None:
        """结束 session：flush + 写 endedAt。"""
        await self.flush_to_db(interview_id)
        async with async_session_factory() as session:
            row = await session.get(SessionCost, interview_id)
            if row:
                from datetime import datetime, timezone
                row.ended_at = datetime.now(timezone.utc)
                await session.commit()

    async def flush_to_db(self, interview_id: str) -> None:
        """强制 flush：把 Redis counter 刷到 SessionCost 表。"""
        if not interview_id or interview_id in ("unknown", "anonymous"):
            logger.warning(f"[FLUSH] skip invalid interviewId={interview_id}")
            return

        key = self._redis_key(interview_id)
        redis = get_redis()
        raw = await redis.hgetall(key)
        if not raw:
            return

        # 计算 cost（按 provider 单价）
        input_price = float(os.getenv("QWEN_INPUT_PRICE", "0.004"))
        output_price = float(os.getenv("QWEN_OUTPUT_PRICE", "0.012"))
        cache_discount = 0.4

        total_prompt = int(raw.get("totalPromptTokens", 0) or 0)
        total_completion = int(raw.get("totalCompletionTokens", 0) or 0)
        cached_tokens = int(raw.get("cachedTokens", 0) or 0)
        uncached_prompt = total_prompt - cached_tokens
        cost = (
            (uncached_prompt / 1000) * input_price
            + (cached_tokens / 1000) * input_price * cache_discount
            + (total_completion / 1000) * output_price
        )

        async with async_session_factory() as session:
            # upsert by interview_id
            stmt = select(SessionCost).where(SessionCost.interview_id == interview_id)
            result = await session.execute(stmt)
            row = result.scalar_one_or_none()
            data = {
                "llm_calls": int(raw.get("llmCalls", 0) or 0),
                "total_prompt_tokens": total_prompt,
                "total_completion_tokens": total_completion,
                "total_tokens": total_prompt + total_completion,
                "prompt_cache_hits": int(raw.get("promptCacheHits", 0) or 0),
                "prompt_cache_misses": int(raw.get("promptCacheMisses", 0) or 0),
                "cached_tokens": cached_tokens,
                "semantic_cache_hits": int(raw.get("semanticCacheHits", 0) or 0),
                "semantic_cache_misses": int(raw.get("semanticCacheMisses", 0) or 0),
                "cache_saved_tokens": int(raw.get("cacheSavedTokens", 0) or 0),
                "retries": int(raw.get("retries", 0) or 0),
                "fallbacks": int(raw.get("fallbacks", 0) or 0),
                "errors": int(raw.get("errors", 0) or 0),
                "input_cost_per_1k": input_price,
                "output_cost_per_1k": output_price,
                "cache_discount": cache_discount,
                "estimated_cost_cny": cost,
            }
            if row:
                for k, v in data.items():
                    setattr(row, k, v)
            else:
                import secrets
                row = SessionCost(
                    id=f"c{secrets.token_hex(12)}",
                    interview_id=interview_id,
                    **data,
                )
                session.add(row)
            await session.commit()
        logger.warning(f"[FLUSH] OK interviewId={interview_id}")

    async def get_cost_panel(self, interview_id: str) -> dict:
        """GET /api/session/:id/cost 用的快路径。"""
        await self.flush_to_db(interview_id)
        async with async_session_factory() as session:
            stmt = select(SessionCost).where(SessionCost.interview_id == interview_id)
            result = await session.execute(stmt)
            row = result.scalar_one_or_none()
            if not row:
                return {
                    "sessionId": interview_id,
                    "totalTokens": 0, "llmCalls": 0,
                    "promptCacheHits": 0, "promptCacheMisses": 0, "promptCacheHitRate": 0,
                    "semanticCacheHits": 0, "semanticCacheMisses": 0, "semanticCacheHitRate": 0,
                    "cacheSavedTokens": 0,
                    "retryRate": 0, "fallbackRate": 0,
                    "estimatedCostCny": 0, "durationMs": 0,
                }
            prompt_total = (row.prompt_cache_hits or 0) + (row.prompt_cache_misses or 0)
            semantic_total = (row.semantic_cache_hits or 0) + (row.semantic_cache_misses or 0)
            from datetime import datetime
            now_naive = datetime.utcnow()
            duration_ms = (
                int((row.ended_at - row.started_at).total_seconds() * 1000)
                if row.ended_at
                else int((now_naive - row.started_at).total_seconds() * 1000)
            )
            return {
                "sessionId": interview_id,
                "totalTokens": row.total_tokens or 0,
                "llmCalls": row.llm_calls or 0,
                "promptCacheHits": row.prompt_cache_hits or 0,
                "promptCacheMisses": row.prompt_cache_misses or 0,
                "promptCacheHitRate": round(row.prompt_cache_hits / prompt_total, 4) if prompt_total else 0,
                "semanticCacheHits": row.semantic_cache_hits or 0,
                "semanticCacheMisses": row.semantic_cache_misses or 0,
                "semanticCacheHitRate": round(row.semantic_cache_hits / semantic_total, 4) if semantic_total else 0,
                "cacheSavedTokens": row.cache_saved_tokens or 0,
                "retryRate": round((row.retries or 0) / row.llm_calls, 4) if row.llm_calls else 0,
                "fallbackRate": round((row.fallbacks or 0) / row.llm_calls, 4) if row.llm_calls else 0,
                "errors": row.errors or 0,
                "estimatedCostCny": round(row.estimated_cost_cny or 0, 4),
                "durationMs": duration_ms,
            }


def get_cost_tracker() -> SessionCostTracker:
    return SessionCostTracker.instance()


async def session_cost_controller(interview_id: str) -> dict:
    """与 NestJS session-cost.controller 等价的 FastAPI handler。"""
    return await get_cost_tracker().get_cost_panel(interview_id)